### DISCUSSION TITLE
RFC-001: Cambio en modelado de datos usuarios/clientes

### DISCUSSION THREAD

**Discussion by adrrf** [ispp-2324-ocial/KB#18] [RFCs]
# Summary

Actualmente, no hay diferencia entre usuarios y cliente. Esto es erroneo ya que claramente tenemos dos tipos de usuario y actualmente solo hay uno

# Basic example

Habria que tener un tipo de usuario el cual sea el que usa la app para descubrir eventos (nuestro usuario) y el que la usará para publicitar sus eventos (nuestro cliente)

# Detailed design

Tendríamos que tener dos hijos del actual usuario de django, uno que sea OcialUser (por ejemplo) y otro que sea OcialClient. El de cliente debería tener asociados los planes y los campos necesarios (tipo de local, por ejemplo) etc. El usuario el chat y rating etc.

# Drawbacks

No hay ningun drawback, debería ser así.

# Adoption strategy

Habría que hacerlo cuanto antes.

**Comment by ferferga** [comment]
En Detailed design no explicas nada. Por qué no pueden ser campos diferentes dentro de la propia entidad de usuario? (entre otras preguntas que seguramente más adelante se me ocurran).

Muevo este issue al KB también y convierto en discusión.

**Reply by adrrf** [reply]
básicamente por que necesitamos dos entidades distintas, un usuario no puede tener eventos ni suscripción, un cliente si los tiene que tener.

**Reply by ferferga** [reply]
Pero por qué no podemos tener una sola entidad y esas propiedades ser opcionales?

Es por esto que el RFC necesita de la sección de "Alternatives", no me queda nada claro por qué, como tu afirmas, esta es la única manera factible de hacer el modelado.

**Reply by adrrf** [reply]
por que no son propiedades, son relaciones

**Reply by ferferga** [reply]
Cuando digo propiedades me refiero a los valores de la entidad transformada a la clase dentro de Django.

Las relaciones claro que pueden ser opcionales, tanto en SQL como en cualquier ORM

Lo que necesitáis es:

```
events = models.ManyToManyField(Event, required=False)
```

**Reply by adrrf** [reply]
la cosa es que para el cliente necesitamos bastante más lógica de negocio y creo que va a ser mucho más comodo y sencillo tener tablas separadas para la separacion de responsabilidades

**Reply by adrrf** [reply]
![imagen](https://github.com/ispp-2324-ocial/KB/assets/72525701/0e403f53-1452-465d-be03-c021243de787)
propongo algo así

**Reply by ferferga** [reply]
Esta mañana voy a estar con otras asignaturas, pero te voy escribiendo con lo que se me ocurre ahora mismo:

* ¿Qué necesitamos de cara al frontend? Porque a mí solo se me ocurre el nombre de la suscripción (para mostrarla) y los permisos que dicha suscripción otorga. Es mucho más fácil comprobar si un usuario tiene permisos cuando estos son una lista de enumerados (como la suscripción), en vez de ir comprobando booleano por booleano.
* Como digo, type puede estar dentro de DjangoUser
* Con este modelo, en backend perdéis la potencia de [Django con sus roles y grupos](https://docs.djangoproject.com/en/5.0/topics/auth/#overview) y tenéis que hacerlo todo a mano.
    * Yo crearía grupos de Django por cada tipo de suscripción y roles como permisos. Creo que es posible además anidar grupos (de forma que herede los permisos del grupo superior), pero esto lo tenéis que investigar vosotros.  El nombre del grupo sería el nombre de la suscripción.
    * Desconozco si a esos grupos se le pueden añadir atributos (igual como digo tenéis que investigarlo vosotros), pero el atributo sería ``price`` y ya tendríamos el precio. **En tu modelo no aparece el precio por ninguna parte**.
* ¿Para qué regla de negocio necesitaríamos almacenar la ubicación del usuario? En mi cabeza, al actualizarse la ubicación del usuario, el frontend enviaría las coordenadas y el backend respondería con todos los eventos en un radio de X, no veo la necesidad de almacenar nada y además esto es un dato de carácter sensible.
* ``typeClient`` no puede tener campos ``locationLat`` y ``locationLong``, esto va en contra de la propia definición de un enumerado.
* Un ``OcialUser`` solo puede subir eventos de un solo tipo? Es decir, si me registro como autónomo y tengo una ludoteca, ¿estoy condenado de por vida a hacer eventos para niños (imaginándonos que sería una categoría)? ¿Y si organizo algo deportivo para adolescentes?

Si te hacen falta más detalles, como te digo, esta tarde a ver si te puedo hacer un diagrama.

**Reply by Domi-ATLAS** [reply]
Fer, lo primero es que la separación es necesaria, ya que necesitamos una relación de usuario a rating, y de cliente a evento, si ponemos un atributo diferenciador en ocialUser, estas relaciones no se podrian modelar debidamente, esto ocurre, para mejorar la legibilidad de la estructura, ademas, asi solo tendriamos que comprobar los clientes, y no comprobar tambien usuarios, los cuales son los unicos que tienen suscripciones. En cuanto a lo demas creo que deberia hablarse con detenimiento en alguna reunión, por que creo que nos estamos complicando la vida, con lo sencillo que se puede hacer todo.

**Reply by ferferga** [reply]
@Domi-ATLAS No hay mención a los ratings en la imagen y por eso no me he acordado en considerarlos. Igualmente porque las relaciones que mencionas no pueden ser opcionales como describo [aquí](https://github.com/ispp-2324-ocial/KB/discussions/18#discussioncomment-8590566)?

**Reply by adrrf** [reply]
Yo pìenso en las relaciones, que para el evento tiene que estar relacionado con un cliente, y no con una entidad que englobe a todos los tipos de usuarios, hay que pensar en la separación de responsabilidades de cada entidad.

**Comment by ferferga** [comment]
CC @ispp-2324-ocial/backend

(Para suscribiros automáticamente a este hilo)

**Reply by adrrf** [reply]
En discord he propuesto un modelo, ya que ratings, solo van a poder poner los usuarios no los clientes
